test/e2e/uniswap_match.ts

```diff
@@ -0,0 +1,199 @@
+): Promise<UniswapPairState> {
+  const uniswapPair = new Contract(
+    await uniswapFactory.getPair(sellToken, buyToken),
+    UniswapV2Pair.abi,
+    provider,
+  );
+  const sellTokenContract = new Contract(sellToken, ERC20.abi, provider);
+  const buyTokenContract = new Contract(buyToken, ERC20.abi, provider);
+
+  const [reserveSellToken, reserveBuyToken, pairToken0] = await Promise.all([
+    sellTokenContract.balanceOf(uniswapPair.address),
+    buyTokenContract.balanceOf(uniswapPair.address),
+    uniswapPair.token0(),
+  ]);
+
+  const isSellTokenToken0 = sellToken === pairToken0;
+  console.log(isSellTokenToken0);
+  return {
+    uniswapPair,
+    reserveSellToken,
+    reserveBuyToken,
+    sellToken: sellTokenContract,
+    buyToken: buyTokenContract,
+    isSellTokenToken0,
+  };
+}
+
+/**
+ * A very basic solver that takes the amount a user wishes to trade on Uniswap
+ * and uses it to:
+ * 1. transform the swap into a GPv2 sell order to sign;
+ * 2. given the signed order, settle the trade onchain.
+ *
+ * This class is single use: each new Uniswap order requires a new instance of
+ * this class.
+ */
+export class UniswapMatch {
+  private readonly uniswapFactory: Contract;
+  private readonly settlement: Contract;
+  private readonly domainSeparator: Promise<TypedDataDomain>;
+  private readonly provider: providers.JsonRpcProvider;
+  order?: Order;
+  uniswapState?: UniswapPairState;
+
+  constructor({
+    uniswapFactoryAddress,
+    settlementAddress,
+    provider,
+  }: {
+    uniswapFactoryAddress: string;
+    settlementAddress: string;
+    provider: providers.JsonRpcProvider;
+  }) {
+    this.uniswapFactory = new Contract(
+      uniswapFactoryAddress,
+      UniswapV2Factory.abi,
+      provider,
+    );
+    this.settlement = new Contract(
+      settlementAddress,
+      GPv2Settlement.abi,
+      provider,
+    );
+    this.provider = provider;
+    this.domainSeparator = provider
+      .getNetwork()
+      .then(({ chainId }) => domain(chainId, this.settlement.address));
+  }
+
+  async createSellOrder({
+    sellToken,
+    buyToken,
+    sellAmount,
+  }: UniswapTradeRequest): Promise<Order> {
+    sellAmount = BigNumber.from(sellAmount);
+
+    const uniswapState = await getUniswapState(
+      this.uniswapFactory,
+      {
+        sellToken,
+        buyToken,
+      },
+      this.provider,
+    );
+
+    const buyAmount = uniswapState.reserveBuyToken
+      .mul(sellAmount)
+      .mul(997)
+      .div(uniswapState.reserveSellToken.mul(1000).add(sellAmount.mul(997)));
+
+    this.uniswapState = uniswapState;
+
+    this.order = {
+      kind: OrderKind.BUY,
+      partiallyFillable: false,
+      buyToken,
+      sellToken,
+      buyAmount,
+      sellAmount,
+      feeAmount: 0,
+      validTo: 0xffffffff,
+      appData: 0,
+    };
+
+    return this.order;
+  }
+
+  async submitSolution(
+    signature: SignatureLike,
+    scheme: SigningScheme,
+    solver: Signer,
```

Comment:
nit: I'd consider the solver more part of the instance state that could be specified at construction time.